### ISSUE TITLE
SignalR receiving events every 5 seconds from ReceiveSessionUpdate event

### ISSUE THREAD

**Issue by LeCloutPanda** [Yellow-Dog-Man/Resonite-Issues#2621]
### Describe the bug?

I don't know if this is a bug or what but I wanna say it is but a few months back I noticed session events on trigger every 30 seconds or so and now it's triggering every five seconds, I was about to check but just check the time stamps of a session when it appeared again and I believe this isn't intended.

### To Reproduce

Use SignalR either in game or external and look for the ``ReceiveSessionUpdate`` event and notice how it triggers every 5 seconds.

### Expected behavior

Not to trigger every 5 seconds but instead every 30 seconds to minimize bandwidth use and still have a consistent flow of updates.

### Screenshots

N/A

### Resonite Version Number

2024.7.17.1173

### What Platforms does this occur on?

Windows, Linux

### What headset if any do you use?

Desktop

### Log Files

N/A can't really supply log files if it is a SignalR thing that can be used outside of the platform.

### Additional Context

I used a custom app to read the SignalR events but I don't doubt the same events are going through Resonite when sessions are updated

### Reporters

LeCloutPanda (Discord)

**Comment by Frooxius** [comment]
This seems normal. SignalR and session updates are sent whenever something changes, not at regular intervals, because it's a push notification system.

The more sessions there are and the more people change things, the more updates you'll get. They're relatively small though, so there's not really a big concern. If someone spams things too much, they should get rate limited.

**Comment by LeCloutPanda** [comment]
I understand that but I get all the sessions when it sends the update rather than the sessions that actually updated, when looking at specific sessions between updates they stay the same so something is for sure wrong here and isn't sending the updated sessions and instead is sending all of them again.

**Comment by Frooxius** [comment]
It should only send the session that actually updated. I don't really know what exactly you're doing - you haven't provided any logs or details, just a general verbal description, so there's not much I can do here.

If you can provide any details, I can have a look, but without knowing specifics, I can't do much.

**Comment by Frooxius** [comment]
Did you do any diffs for the received data to see if there's really nothing changing?

The way the system is written, it will only generate updates if it detects that any of the parameters have changed - and then on top of that it'll send update every now and then.

I don't see anything in the data that would indicate that nothing really changed - it's possible that a parameter actually is changing, forcing the updates.

**Comment by LeCloutPanda** [comment]
Just did a diff check and the only variables that were different was the ``lastUpdate`` variable, if that is all that is being changed then this issue might be more then that and something to do with how sessions update needs to be updated.

![image](https://github.com/user-attachments/assets/95e52355-214b-4c56-a774-a409fadbe91e)

**Comment by Frooxius** [comment]
Thanks! Yeah, if nothing else is changing, then there's likely something that's triggering updates when it shouldn't. 

Do you have logs from the host of the session? They can indicate on why is session info being updated.

**Comment by LeCloutPanda** [comment]
Here are the few logs I have, I literally set the headless up just for this so it's brand new
[GEOFF - 2024.7.22.1319 - 2024-07-25 00_00_10.log](https://github.com/user-attachments/files/16363502/GEOFF.-.2024.7.22.1319.-.2024-07-25.00_00_10.log)
[GEOFF - 2024.7.22.1319 - 2024-07-24 23_51_45.log](https://github.com/user-attachments/files/16363504/GEOFF.-.2024.7.22.1319.-.2024-07-24.23_51_45.log)
[GEOFF - 2024.7.22.1319 - 2024-07-25 00_14_36.log](https://github.com/user-attachments/files/16363505/GEOFF.-.2024.7.22.1319.-.2024-07-25.00_14_36.log)

**Comment by LeCloutPanda** [comment]
With the 2025.1.25.1219 update does that fix this issue as well?

**Comment by Frooxius** [comment]
No.

**Comment by shiftyscales** [comment]
This sounds like it's also potentially related to #2970.

**Comment by Nutcake** [comment]
To add another datapoint to this, with a simple RML mod to just print when `ReceiveSessionUpdate` gets called, as well as when I connect to the SignalR hub with a non-Resonite app, I get on average about 19 `ReceiveSessionUpdate` calls _every second_ .

This is a log with 4362 calls of `ReceiveSessionUpdate` over a time period of about 4 minutes.
[Resonite.log](https://github.com/user-attachments/files/18595732/Resonite.log)

```csharp
namespace RSULog
{
    public class RSULog : ResoniteMod
    {
        public override string Name => "RSU Log";
        public override string Author => "Nutcake";
        public override string Version => "1.0.0";

        public override void OnEngineInit()
        {
            var harmony = new Harmony("rsulog");
            harmony.PatchAll();
        }

        [HarmonyPatch(typeof(HubStatusController), methodName: "ReceiveSessionUpdate")]
        class ReceiveSessionUpdatePatch
        {
            public static void Prefix(HubStatusController __instance)
            {
                Msg("ReceiveSessionUpdate called");
            }
        }
    }
}
```

A lot of these session updates only differ in their `lastUpdate` time with the time values about 5 seconds apart.

**Comment by shiftyscales** [comment]
It sounds like we have a likely known cause for what is leading to this / this could possibly be a quick fix (or something that is quick to look into / dismiss if it turns out to be unrelated.)

Curious for input from @BlueCyro who is currently assigned to related issue #2970.

This is a nice one to look into / fix as it should make SignalR more efficient so that it actually only pushes when a meaningful change is made

**Comment by Frooxius** [comment]
If this was a simple fix, we'd have fixed this already, there's something more complicated going on there that needs to be investigated. 

We have a mechanism that should prevent the changes from being sent when nothing is changed, but it's more fuzzy in some things, so it might need some restructuring.

This isn't really related to #2970, that's more of a symptom/consequence of this, but a separate issue on its own.

**Comment by shiftyscales** [comment]
As above though- is having a different `lastUpdate` timestamp irrelevant then @Frooxius? I haven't seen any subsequent updates to this issue to indicate that lead has been debunked.

Based on the prior findings by @LeCloutPanda and the above findings by @Nutcake - it seems that is the only property that is changing that consistently.

**Comment by Frooxius** [comment]
@shiftyscales Probably? I don't know, this needs to be investigated to answer those questions.

**Comment by XDelta** [comment]
I was also able to replicate @Nutcake's findings seeing ~20 `ReceiveSessionUpdate` calls every second on average (Amount received varies on how many sessions are running). When narrowing it down to looking at just one session, I was getting an update every ~5s and diffing the response, pretty much all the updates only differed on `lastUpdate` time having changed.

**Comment by eth0fox** [comment]
yeah i've been concerned about this, primarily for network bandwidth reasons - i had a connection open for just over 45 minutes and in that time recieved about 165MB worth of incoming websocket traffic according to Firefox' network debugger, so about a constant 460kbps just keeping the session open (this was at around 2300 UTC, I'd assume it'd be slightly higher at peak times).

It's probably not a critical issue for real players (i assume session traffic is much more), but while I understand this isn't going to be a huge priority, for users of light clients like ReCon, who might be on slower or metered connections, (I've put some devtime into my chat client while on a train and it's less than stellar). And of course, if you extrapolate to the scale of the entire platform I don't know how happy those numbers are, but that's not my place to talk I realise

<img width="569" height="1245" alt="Image" src="https://github.com/user-attachments/assets/0d41b8d6-d686-47e1-b37c-84f4416e9e9a" />
